### PR TITLE
Draft architecture docs

### PR DESCRIPTION
**Pull request by pauldougan**
Initial draft of some high level documentation for the GSP

### CODE REVIEW

File: docs/gsp-architecture-overview.md [hunk 1]
```diff
@@ -0,0 +1,53 @@
+#  GDS Supported Platform Overview
+
+## Generic View
+
+This view shows the high level componemts  that make up the GDS Supported Platform.
```
Comment (pauldougan) [inline docs/gsp-architecture-overview.md]: typo components

File: README.md [hunk 2]
```diff
@@ -1,8 +1,36 @@
...
+
+## What do we currently support?
+
+- A declarative continuous delivery workflow - merging to master triggers deployment to production
+- A container platform based on industry standard [Docker](https://docs.docker.com/) and [Kubernetes](https://kubernetes.io)
+- Build tooling with [ConcourseCI](https://concourse-ci.org/)
+- A private container registry with [Docker Registry](https://docs.docker.com/registry/)
+- Sign and verify docker image integrity with [Docker Notary](https://docs.docker.com/notary/)
+- Scanning of docker images for security vulnerabilities with [clair](https://github.com/coreos/clair)
+- Applicaton deployment with [Flux](https://github.com/weaveworks/flux)
```
Comment (jonathanglassman) [inline README.md]: Applicaton => Application

File: README.md [hunk 3]
```diff
@@ -1,8 +1,36 @@
...
+
+The platform is for teams working in the [Government Digital Service](https://www.gov.uk/government/organisations/government-digital-service) that need to run software applications.
+
+## What do we currently support?
+
+- A declarative continuous delivery workflow - merging to master triggers deployment to production
+- A container platform based on industry standard [Docker](https://docs.docker.com/) and [Kubernetes](https://kubernetes.io)
+- Build tooling with [ConcourseCI](https://concourse-ci.org/)
+- A private container registry with [Docker Registry](https://docs.docker.com/registry/)
+- Sign and verify docker image integrity with [Docker Notary](https://docs.docker.com/notary/)
```
Comment (jonathanglassman) [inline README.md]: "Signing and verification..." to make consistent with other bullets

File: README.md [hunk 4]
```diff
@@ -1,8 +1,36 @@
...
+## Who is the platform for?
+
+The platform is for teams working in the [Government Digital Service](https://www.gov.uk/government/organisations/government-digital-service) that need to run software applications.
+
+## What do we currently support?
+
+- A declarative continuous delivery workflow - merging to master triggers deployment to production
+- A container platform based on industry standard [Docker](https://docs.docker.com/) and [Kubernetes](https://kubernetes.io)
+- Build tooling with [ConcourseCI](https://concourse-ci.org/)
+- A private container registry with [Docker Registry](https://docs.docker.com/registry/)
```
Comment (blairboy362) [inline README.md]: Harbor is the registry - it implements the docker registry API.

File: README.md [hunk 5]
```diff
@@ -1,8 +1,36 @@
...
+- Sign and verify docker image integrity with [Docker Notary](https://docs.docker.com/notary/)
+- Scanning of docker images for security vulnerabilities with [clair](https://github.com/coreos/clair)
+- Applicaton deployment with [Flux](https://github.com/weaveworks/flux)
+- Monitoring and alerting with [Prometheus](https://prometheus.io/), [Alertmanager](https://prometheus.io/docs/alerting/alertmanager/) and [Grafana](https://grafana.com/)
+- Secrets management with [sealed-secrets](https://github.com/bitnami-labs/sealed-secrets)
```
Comment (blairboy362) [inline README.md]: I would argue the "management" of the secrets is handled by kubernetes; "secure, automated injection" is the part provided by "sealed-secrets" (in concert with the flux helm operator).

File: README.md (hunk 3 above)
Comment (blairboy362) [reply README.md]: Depending on what we mean by "current" in the heading of this section...we don't do the "verification" step yet. It's currently in the plan, but we're not there yet.

File: docs/gsp-architecture-continuous-deployment.md [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+
+<!--__[edit draw.io diagram](https://www.draw.io/?state=%7B%22ids%22:%5B%221p4qkP2-fsMpc42ZCsKIVlGwJ4blmnu_q%22%5D,%22action%22:%22open%22,%22userId%22:%22104206899246339571570%22%7D#G1p4qkP2-fsMpc42ZCsKIVlGwJ4blmnu_q)__
+-->
+
+![GDS Supported Platform Continuous Delivery](diagrams/gsp-architecture-continuous-delivery.svg)
+
+|Seq|Step|Description|
+|-|-|-|
+|1.|Raise pull request|The Developer raises a pull request to review changes to the source code repository|
+|2.|Merge pull request to master|A second developer reviews and approves the changes in the pull request and merges to the master branch|
```
Comment (blairboy362) [inline docs/gsp-architecture-continuous-deployment.md]: It requires two other developers (three total, including the author of the PR).

File: docs/gsp-architecture-continuous-deployment.md [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+|3.|Check change to master|Determine if the master branch has changed and needs building|
+|4.|Check for 2-eyes| Ensure that the change has been signed by two different developers|
+|5.|Check signers on approved list|Check the change has been signed by approved developers|
+|6.|Trigger build task|Start the pipeline to build from the master branch|
+|7.|Build images|Build the Docker images|
+|8.|Run unit tests|Unit test the Docker images|
+|9.|Push images |Push successfully built Docker images to the Docker Registry|
+|10.|Tag and sign Docker images|Tag the newly built Docker images and sign them to enable the checking of provenance during deployent|
+|11.|Bump image tag|Increment the tag on the Docker image|
```
Comment (blairboy362) [inline docs/gsp-architecture-continuous-deployment.md]: Currently this is a manual step carried out by the developers; and the flow here may change before we get to implement it.

File: docs/gsp-architecture-continuous-deployment.md [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+|12.|Check for deploy|Determine if a deployment needs to be applied to the cluster|
+|13.|Trigger deploy|Start the deployment|
+|14.|Check deployment provenance|Check that the deployment is trusted and from a known source|
```
Comment (blairboy362) [inline docs/gsp-architecture-continuous-deployment.md]: I'm pretty sure we don't do this and not sure when we do the piece of work that delivers this. We may not want to shout about it.

File: docs/gsp-architecture-continuous-deployment.md [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+|15.|Render deploy template |Merge the secrets and variables with the deployment template|
+|16.|Apply deploy|Apply the deployment to the cluster|
+|17.|Pull images|Pull images referred to in the cluster from the private Docker Regsitry|
+|18.|Return image|Registry returnd the Docker image|
+|19.|Wait for healthy state|Wait until the deployment has been applied to the cluster and the deployed application is in a healthy state|
+|20.|Run acceptance tests|Run any post installation tests|
+|21.|Tag and sign deployment|Label the successfully applied deployment with a tag and sign|
```
Comment (blairboy362) [inline docs/gsp-architecture-continuous-deployment.md]: We should be clear here that we mean "tag/sign to mark it OK for promotion to next step (e.g. production)".

File: docs/gsp-architecture-infrastructure.md [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+-->
+
+
+The cluster sits in a single AWS account within a single region (__eu-west-2__ - London)
+
+The infrastructure is split across the three availability zones __eu-west-2a__, __eu-west-2b__, __eu-west-2c__
+
+The state for the cluster is persisted across three etcd nodes in different availability zones.
+
+By default there are two masters for availability and resilience but this can be overridden.
```
Comment (blairboy362) [inline docs/gsp-architecture-infrastructure.md]: three masters is the default, now....one per AZ.

Comment (pauldougan) [reply docs/gsp-architecture-infrastructure.md]: updated the text and the diagram

File: README.md (hunk 4 above)
Comment (chrisfarms) [reply README.md]: AFAICT harbor "registry" is just the docker distribution "registry" but built on the same "photon" base image as all the others ... I think it's fine to not call out Harbor specifically unless we want to start letting people use the Harbor "portal" for managing users/repositories

File: README.md (hunk 5 above)
Comment (chrisfarms) [reply README.md]: maybe "Secure git-based secrets configuration via sealed-secrets" ?

File: docs/gsp-architecture-continuous-deployment.md (hunk 6 above)
Comment (chrisfarms) [reply docs/gsp-architecture-continuous-deployment.md]: three is an annoying artifact of our implementation I think... the aim/requirement is minimum two people total have contributing to getting a change merged

File: README.md (hunk 4 above)
Comment (pauldougan) [reply README.md]: Chris, I was about to ask you about this but you beat me to it. I'll stick with the more generic Docker Registry.

File: README.md (hunk 5 above)
Comment (pauldougan) [reply README.md]: I'll go with Secure git-based secrets configuration via sealed-secrets

File: docs/gsp-architecture-continuous-deployment.md (hunk 6 above)
Comment (pauldougan) [reply docs/gsp-architecture-continuous-deployment.md]: how about this @chrisfarms  

"Require a minimum of two developers to approve the change before allowing the merge to the master branch"

File: docs/gsp-architecture-continuous-deployment.md (hunk 9 above)
Comment (pauldougan) [reply docs/gsp-architecture-continuous-deployment.md]: done

File: README.md (hunk 3 above)
Comment (pauldougan) [reply README.md]: is this section better described as features?

Comment (pauldougan) [reply README.md]: ok its changed

**Pull request merged**